src/cargo/util/network.rs

```diff
@@ -1,36 +1,62 @@
 use anyhow::Error;
+use std::cmp::min;
+use std::thread::sleep;
+use std::time::Duration;
+
+use log::debug;
 
 use crate::util::errors::{CargoResult, HttpNot200};
 use crate::util::Config;
 
 pub struct Retry<'a> {
     config: &'a Config,
-    remaining: u32,
+    retries: u32,
+    max_retry: u32,
+    retry_max_time: Duration,
+    retry_delay: Option<Duration>,
 }
 
 impl<'a> Retry<'a> {
     pub fn new(config: &'a Config) -> CargoResult<Retry<'a>> {
         Ok(Retry {
             config,
-            remaining: config.net_config()?.retry.unwrap_or(2),
+            max_retry: config.net_config()?.retry.unwrap_or(2),
+            retries: 0,
+            retry_max_time: config
+                .net_config()?
+                .retry_max_time
+                .map_or(Duration::from_secs(32), |e| e.inner()),
+            retry_delay: config.net_config()?.retry_delay.map(|e| e.inner()),
         })
     }
 
     pub fn r#try<T>(&mut self, f: impl FnOnce() -> CargoResult<T>) -> CargoResult<Option<T>> {
         match f() {
-            Err(ref e) if maybe_spurious(e) && self.remaining > 0 => {
+            Err(ref e) if maybe_spurious(e) && self.retries < self.max_retry => {
                 let msg = format!(
                     "spurious network error ({} tries remaining): {}",
-                    self.remaining,
+                    (self.max_retry - self.retries),
                     e.root_cause(),
                 );
                 self.config.shell().warn(msg)?;
-                self.remaining -= 1;
+                self.retries += 1;
+                self.backoff();
                 Ok(None)
             }
             other => other.map(Some),
         }
     }
+
+    fn backoff(&self) {
+        let backoff_time = if let Some(delay) = self.retry_delay {
+            delay
+        } else {
+            min(Duration::from_secs(1 << self.retries), self.retry_max_time)
```

Comment:
Not sure if we want to be too granular for exponential backoff time. I don't see this being used too often

Reply:
Can you elaborate on what you mean by that? Do you fear that this is too many knobs to expose through configuration?

Reply:
> Do you fear that this is too many knobs to expose through configuration?

Yes